ipld/resolver/src/behaviour/membership.rs

```diff
@@ -0,0 +1,241 @@
+    }
+
+    /// Mark a peer as routable in the cache.
+    ///
+    /// Call this method when the discovery service learns the address of a peer.
+    pub fn set_routable(&mut self, peer_id: PeerId) {
+        self.provider_cache.set_routable(peer_id)
+    }
+
+    /// Mark a peer as unroutable in the cache.
+    ///
+    /// Call this method when the discovery service forgets the address of a peer.
+    pub fn set_unroutable(&mut self, peer_id: PeerId) {
+        self.provider_cache.set_unroutable(peer_id);
+        self.outbox.push_back(Event::Removed(peer_id))
+    }
+
+    /// List the current providers of a subnet.
+    ///
+    /// Call this method when looking for a peer to resolve content from.
+    pub fn providers_of_subnet(&self, subnet_id: &SubnetID) -> Vec<PeerId> {
+        self.provider_cache.providers_of_subnet(subnet_id)
+    }
+
+    /// Parse and handle a [`GossipsubMessage`]. If it's from the expected topic,
+    /// then raise domain event to let the rest of the application know about a
+    /// provider. Also update all the book keeping in the behaviour that we use
+    /// to answer future queries about the topic.
+    fn handle_message(&mut self, msg: GossipsubMessage) -> Option<Event> {
+        if msg.topic == self.membership_topic.hash() {
+            match SignedProviderRecord::from_bytes(&msg.data).map(|r| r.into_record()) {
+                Ok(record) => match self.provider_cache.add_provider(&record) {
+                    None => return Some(Event::Skipped(record.peer_id)),
+                    Some(d) if d.is_empty() => return None,
+                    Some(d) => return Some(Event::Updated((record.peer_id, d))),
+                },
+                Err(e) => {
+                    warn!(
+                        "Gossip message from peer {:?} could not be deserialized: {e}",
+                        msg.source
+                    );
+                }
+            }
+        } else {
+            warn!("unknown gossipsub topic: {}", msg.topic);
+        }
+        None
+    }
+
+    /// Remove any membership record that hasn't been updated for a long time.
+    fn prune_membership(&mut self) {
+        let cutoff_timestamp = Timestamp::now() - self.max_provider_age;
+        let pruned = self.provider_cache.prune_providers(cutoff_timestamp);
+        for peer_id in pruned {
+            self.outbox.push_back(Event::Removed(peer_id))
+        }
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ConnectionHandler = <Gossipsub as NetworkBehaviour>::ConnectionHandler;
+    type OutEvent = Event;
+
+    fn new_handler(&mut self) -> Self::ConnectionHandler {
+        self.inner.new_handler()
+    }
+
+    fn addresses_of_peer(&mut self, peer_id: &PeerId) -> Vec<Multiaddr> {
+        self.inner.addresses_of_peer(peer_id)
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        self.inner.on_swarm_event(event)
+    }
+
+    fn on_connection_handler_event(
```

Comment:
Do we want to already add a pperID to the routable list when we establish a connection to it? (you may be addressing this in a folow up PR, let me know if this is the case, but I didn't want the idea to be lost).

Reply:
Right, you are saying that Kademlia knowing the address in the k-bucket of a peer is one thing, but a peer being connected to us is another. 

It depends on whether `Gossipsub::addresses_of_peers` would return the address of the peer or not. Looking at the source it doesn't return anything, so we would only be able to route to the peer if we remembered its address on our own, or if Kademlia does it for us. 

By the way I just saw that this changed. In [v0.50.0](https://github.com/libp2p/rust-libp2p/blob/v0.50.0/swarm/src/lib.rs#L569) it called `addresses_of_peers` but not in the latest version.